Cherry-pick changes for ComputeClangResourceDirectory on Windows

bulbazord opened the pull request:
This is useful for having swift debugging on a Windows host, so I'd like to cherry-pick them into stable.

cc @compnerd @adrian-prantl

bulbazord committed:
[build.py] Allow clang-cl to build files starting with '/U'

Summary:
clang-cl tries to match cl's interface, and treats /U as "Removes a
predefined macro" as cl does. When you feed clang-cl a file that begins with
'/U' (e.g. /Users/xiaobai/foo.c), clang-cl will emit a warning and in some cases
an error, like so:

clang-9: warning: '/Users/xiaobai/foo.c' treated as the '/U' option [-Wslash-u-filename]
clang-9: note: Use '--' to treat subsequent arguments as filenames
clang-9: error: no input files

If you're using clang-cl, make sure '--' is passed before the source file.

Differential Revision: https://reviews.llvm.org/D58860

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355341 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry-picked from commit 11b8388ae5ec73c53bf5f3bb1101c03a583f7de1)

bulbazord committed:
[ExpressionParser] Test GetClangResourceDir

Summary:
I'm doing this because I plan on implementing `ComputeClangResourceDirectory`
on windows so that `GetClangResourceDir` will work.  Additionally, I made
test_paths make sure that the directory member of the returned FileSpec is not
none. This will fail on windows since `ComputeClangResourceDirectory` isn't
implemented yet.

Differential Revision: https://reviews.llvm.org/D58748

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355463 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry-picked from commit 8976af1d17f905f29d70e0caf8e221b9d0a9fdc7)

bulbazord committed:
[ExpressionParser] Fix ComputeClangResourceDirectory for windows

The function signature of ComputeClangResourceDirectory for windows
wasn't updated when the others changed, causing the windows build to
fail. This should fix that.

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355471 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry-picked from commit 621e35c002fff14eb156ad866ce0499787882dac)

bulbazord committed:
[ExpressionParser] Implement ComputeClangResourceDir for Windows

Summary: This function is useful for expression evaluation, especially when doing swift debugging on windows.

Reviewers: aprantl, labath

Reviewed By: labath

Subscribers: teemperor, jdoerfert, lldb-commits

Tags: #lldb

Differential Revision: https://reviews.llvm.org/D59072

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355631 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry-picked from commit 73f4149a470633fe26acbb51f89fac50722fea2b)

bulbazord committed:
Fix TestPaths.py on windows

I committed an implementation of GetClangResourceDir on windows but
forgot to update this test. I merged the tests like I intended to, but I
realized that the test was actually failing. After looking into it, it
appears that FileSystem::Resolve was taking the path and setting
the FileSpec's Directory to "/path/to/lldb/lib/clang/" and the File to
"9.0.0" which isn't what we want. So I removed the resolve line from
DefaultComputeClangResourceDir.

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355648 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry-picked from commit b3fe704527ae502410fea4e189d1df3d923a0a5f)

bulbazord committed:
[ExpressionParser] Adjust default clang resource dir for swift

The previous patch refactored DefaultClangResourceDir and swift-lldb
puts the clang resource dir in a different location.

bulbazord committed:
[ExpressionParser] Adjust ClangParserTest to account for swift-lldb differences

compnerd commented on source/Plugins/ExpressionParser/Clang/ClangHost.cpp:
```diff
@@ -53,8 +42,37 @@ static bool VerifyClangPath(const llvm::Twine &clang_path) {
+    return true;
+  }
+
+  return HostInfoPosix::ComputePathRelativeToLibrary(file_spec, relative_path);
```
Why `HostInfoPosix` here?  Should this be conditional on `LLVM_ON_UNIX` and `LLVM_ON_WIN32` perhaps?

bulbazord replied:
I think you commented on one of the commits in the middle? It's definitely not HostInfoPosix in the final result.

compnerd commented on source/Plugins/ExpressionParser/Clang/ClangHost.cpp:
```diff
@@ -67,8 +85,10 @@ bool lldb_private::ComputeClangDirectory(FileSpec &lldb_shlib_spec,
     ++rev_it;
   }
 
+  // Posix-style of LLDB detected.
```
I don't understand this comment.

bulbazord replied:
Yeah looking back this comment isn't very good. I intended it to indicate we aren't in a framework, but rather a more traditional tree with a `bin/ include/ lib/ share/` set of directories. I can update the comment on llvm.org and cherry-pick it into this PR with the rest of these commits.

compnerd commented:
@swift-ci please test

bulbazord committed:
Rewrite comment to be clearer

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355875 91177308-0d34-0410-b5e6-96231b3b80d8

compnerd commented:
@swift-ci please test

compnerd commented:
CC: @dcci @adrian-prantl @fredriss

adrian-prantl committed:
Update the macOS implementation of ComputeClangResourceDirectory to
comply with the more pedantic TestPaths.py

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@355994 91177308-0d34-0410-b5e6-96231b3b80d8

adrian-prantl committed:
Make sure FileSystem::Resolve preserves the path/file distinction.

This should finally fix TestPaths.py.

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@356057 91177308-0d34-0410-b5e6-96231b3b80d8

bulbazord committed:
[ExpressionParser] Restore removed intance of FileSystem::Resolve while resolving clang resource dir

I committed a change that removed this line, but I meant to restore it
befor committing.

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@356079 91177308-0d34-0410-b5e6-96231b3b80d8

bulbazord commented:
ping! @adrian-prantl

adrian-prantl commented:
Can you also add r356057 and r355994? I had to fixup the macOS implementation a little bit to get all tests to pass.

adrian-prantl commented:
Ah, you did!

compnerd commented:
@swift-ci please test

adrian-prantl commented on source/Plugins/ExpressionParser/Clang/ClangHost.cpp:
```diff
@@ -53,9 +36,36 @@ static bool VerifyClangPath(const llvm::Twine &clang_path) {
+
+  llvm::SmallString<256> clang_dir(parent_dir);
+  llvm::SmallString<32> relative_path;
+  llvm::sys::path::append(relative_path, "lib", "lldb", "clang");
```
@xiaobai This is now different from llvm.org, where this line reads:
```
  llvm::sys::path::append(relative_path,
                          llvm::Twine("lib") + CLANG_LIBDIR_SUFFIX, "clang",
                          CLANG_VERSION_STRING);
```

Why is that? And could you please make sure to bring both versions back in sync?

bulbazord replied:
From what I understand, swift-lldb relies on the clang resource copied from swift, which is placed in `$install_location/lib/lldb/clang`, whereas llvm.org's lldb relies on the clang resource dir located at `$install_location/lib/clang/$clang_version`. I'm not sure how I'd synchronize llvm.org and swift-lldb right now. Maybe a cmake variable that fills out Config.h?

adrian-prantl replied:
I think you should be able to do something similar to what the macOS implementation does and just have a list of locations that are searched one after another. You can probably even reuse some of the code in the macOS implementation. If you think that that is too inefficient then we should detect the build type (e.g., by having CMake generate a macro) and make the code depend on that. Depending on the repository/branch is not a good idea, because some people *will* be building the swift-lldb compiler without swift.

bulbazord replied:
Ah yeah looking at what the MacOS implementation does, that sounds like the best option. I should have some time in the next few days to fix that up. Thanks for pointing this out

adrian-prantl replied:
Thanks! These kinds of differences between upstream and swift-lldb often cause extremely difficult to debug problems, especially when related to search paths, so this will pay off quite a lot.

bulbazord replied:
@adrian-prantl I opened #1411 to cherry-pick the change to stable.
